File: src/main/java/Q85_MaximalRectangle.java
import java.util.LinkedList;

/**
 * 1hr 6mins
 * Runtime: 9 ms, faster than 39.38% of Java online submissions for Maximal Rectangle.
 * Memory Usage: 43.3 MB, less than 23.17% of Java online submissions for Maximal Rectangle.
 * Brutal force
 *
 * the best solution is to merge matrix as histogram, then use O(n) histogram algorithm to calculate row by row
 * overall is O(n^2)
 */
public class Q85_MaximalRectangle {
    char[][] matrix;
    public int maximalRectangle(char[][] matrix) {
        if(matrix.length == 0 || matrix[0].length == 0) {
            return 0;
        }
        int result = 0;
        this.matrix = matrix;
        for(int i = 0; i < matrix.length; i++) {
            for(int j = 0; j < matrix[0].length; j++) {
                result = Math.max(result, getMaxSize(i, j));
            }
        }
        return result;
    }

    public int getMaxSize(int i, int j) {
        if(matrix[i][j] == '0') {
            return 0;
        }
        int iEnd = i;
        int jEnd = j;
        int result = 0;
        while(iEnd < matrix.length && matrix[iEnd][j] == '1') {
            iEnd++;
        }
        while(jEnd < matrix[0].length && matrix[i][jEnd] == '1') {
            jEnd++;
        }
        for(int x = i; x < iEnd; x++) {
            for(int y = j; y < jEnd; y++ ) {
                if(matrix[x][y] == '0') {
                    jEnd = y;
                }
                else {
                    result = Math.max((x - i + 1) * (y - j + 1), result);
                }
            }
        }
        return result;
    }
}

File: LeetCodePractice/LeetCodePractice/Q85_MaximalRectangle.cpp
#include<vector>
using namespace std;

class Solution {//4:20 4:27
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        if (matrix.size() == 0 || matrix[0].size() == 0) {
            return 0;
        }
        int m = matrix.size();
        int n = matrix[0].size();
        vector<vector<int>> memo(m, vector<int>(n, 0));
        for (int i = 0; i < m; i++) {
            int count = 0;
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == '1') {
                    count++;
                }
                else {
                    count = 0;
                }
                memo[i][j] = count;
            }
        }
        int result = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] != '1') {
                    continue;
                }
                int col = 1;
                int row = INT_MAX;
                for (int k = i; k >= 0; k--) {
                    if (matrix[k][j] == '0') {
                        break;
                    }
                    row = std::min(row, memo[k][j]);
                    result = std::max(result, row * col);
                    col++;
                }
            }
        }
        return result;
    }
};